File: DynamicProgramming/64_MinimumPathSum.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-


class Solution(object):
    def minPathSum(self, grid):
        if not grid:
            return 0

        m_row = len(grid)
        n_col = len(grid[0])

        # dp[i][j]: the min sum along path from top left to [i,j].
        dp = [[0 for i in range(n_col)] for j in range(m_row)]

        for row in range(m_row):
            for col in range(n_col):
                dp[row][col] = grid[row][col]
                if row == 0 and col > 0:
                    dp[row][col] += dp[row][col - 1]
                elif row > 0 and col == 0:
                    dp[row][col] += dp[row - 1][col]
                elif row > 0 and col > 0:
                    dp[row][col] += min(dp[row - 1][col], dp[row][col - 1])
                else:
                    pass

        return dp[m_row - 1][n_col - 1]


"""
[]
[[0]]
[[1,2,4,3,2,1,5],[3,4,1,2,3,5,4],[3,2,4,5,1,2,5]]
"""

File: DynamicProgramming/64_MinimumPathSum.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-06-30 14:48:13
 */

class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        if(grid.size()==0){
            return 0;
        }
        int rows = grid.size();
        int cols = grid[0].size();

        vector<vector<int>> dp(rows, vector<int>(cols, 0));
        for(int i=0; i<rows; i++){
            for(int j=0; j<cols; j++){
                dp[i][j] = grid[i][j];
                if(i==0 && j>0){
                    dp[i][j] += dp[i][j-1];
                }
                else if(j==0 && i>0){
                    dp[i][j] += dp[i-1][j];
                }
                else if(j > 0 && i>0){
                    dp[i][j] += min(dp[i-1][j], dp[i][j-1]);
                }
            }
        }
        return dp[rows-1][cols-1];
    }
};

/*
[]
[[0]]
[[1,2,4,3,2,1,5],[3,4,1,2,3,5,4],[3,2,4,5,1,2,5]]
*/
